### PR TITLE
Avoid broadcasting transactions to peers that send them

### LINKED ISSUE #3757: Keep track of peers broadcasting the transaction back
Right now we keep track of which peers the transaction was broadcasted to.
We could also check which peers broadcasted the transaction back - that would mean they have the transaction in their queue (so the transactions was accepted by the network).

### PR DESCRIPTION
**Pull request by tomusdrw**
Closes #3757 

It's looks a bit ugly to propagate the peer_id to the client and then back to sync, @arkpar any ideas how to improve this?

@rphmeier in light client implementation we cannot easily get `network.peer_session_info(peer_id)` any ideas how to solve this?

### CODE REVIEW

**Comment by rphmeier** [comment]
@tomusdrw Most likely can be done with an extension to the `EventContext` trait, although I'd prefer something like `persistent_id(&PeerId)` as opposed to exposing the full p2p session info at such a high level. Then see the `TxRelay` handler within `sync/api.rs`

File: sync/src/api.rs [hunk 1]
```diff
@@ -344,7 +351,8 @@ struct TxRelay(Arc<BlockChainClient>);
 impl LightHandler for TxRelay {
 	fn on_transactions(&self, ctx: &EventContext, relay: &[::ethcore::transaction::SignedTransaction]) {
 		trace!(target: "les", "Relaying {} transactions from peer {}", relay.len(), ctx.peer());
-		self.0.queue_transactions(relay.iter().map(|tx| ::rlp::encode(tx).to_vec()).collect())
+		// TODO [ToDr] Can we get a peer enode somehow?
+		self.0.queue_transactions(relay.iter().map(|tx| ::rlp::encode(tx).to_vec()).collect(), None)
```
Comment (rphmeier) [inline sync/src/api.rs]: `ctx.persistent_peer_id(&ctx.peer())` once the method I mentioned has been added.

File: ethcore/light/src/net/context.rs [hunk 2]
```diff
@@ -75,6 +82,9 @@ pub trait EventContext {
     /// disconnected/connected peer.
     fn peer(&self) -> PeerId;
 
+	/// Returns the relevant's peer persistent Id (aka NodeId).
+	fn persistent_peer_id(&self) -> Option<NodeId>;
```
Comment (rphmeier) [inline ethcore/light/src/net/context.rs]: would prefer for this to take arbitrary `&PeerId` ideally.

File: ethcore/src/client/client.rs [hunk 3]
```diff
@@ -559,11 +559,16 @@ impl Client {
...
+	pub fn import_queued_transactions(&self, transactions: &[Bytes], peer_id: Option<H512>) -> usize {
 		trace!(target: "external_tx", "Importing queued");
 		let _timer = PerfTimer::new("import_queued_transactions");
 		self.queue_transactions.fetch_sub(transactions.len(), AtomicOrdering::SeqCst);
-		let txs = transactions.iter().filter_map(|bytes| UntrustedRlp::new(bytes).as_val().ok()).collect();
+		let txs: Vec<SignedTransaction> = transactions.iter().filter_map(|bytes| UntrustedRlp::new(bytes).as_val().ok()).collect();
+		let hashes: Vec<_> = txs.iter().map(|tx| tx.hash()).collect();
+		let block_number = self.chain_info().best_block_number;
+		self.notify(|notify| {
+			notify.transactions_imported(hashes.clone(), peer_id.clone(), block_number);
```
Comment (arkpar) [inline ethcore/src/client/client.rs]: Should not it happen only after successful import?

File: ethcore/light/src/net/context.rs [hunk 4]
```diff
@@ -89,19 +99,26 @@ pub trait EventContext {
...
 impl<'a> EventContext for Ctx<'a> {
-    fn peer(&self) -> PeerId { self.peer }
+
+    fn peer(&self) -> PeerId {
+		self.peer
+	}
+
+	fn persistent_peer_id(&self) -> Option<NodeId> {
+		self.io.persistent_peer_id(self.peer)
+	}
```
Comment (arkpar) [inline ethcore/light/src/net/context.rs]: Indentation looks strange around this block. Does the file use spaces?

File: ethcore/light/src/net/tests/mod.rs [hunk 5]
```diff
@@ -69,7 +69,7 @@ impl IoContext for Expect {
 		Some(super::MAX_PROTOCOL_VERSION)
 	}
 
-	fn persistent_peer_id(&self, _peer: PeerId) -> Option<NodeId> {
+	fn persistent_peer_id(&self, _peer: &PeerId) -> Option<NodeId> {
```
Comment (rphmeier) [inline ethcore/light/src/net/tests/mod.rs]: all other functions of this type take by-value peer id. why make the change?

**Pull request merged**